openprescribing/frontend/management/commands/infer_practice_boundaries.py

```diff
@@ -0,0 +1,70 @@
+"""
+Infer boundaries for practices from their locations using a Voronoi partition
+combined with CCG boundary data.
+
+These boundaries have no real meaning but they're easy enough to explain (we
+assign every point to its closest practice) and they allow us to render
+plausible-looking maps for arbitrary collections of practices.
+
+Practice boundaries are clipped to their CCG boundaries so we can then use the
+simple rule that the boundary of any organisation is the union of the
+boundaries of its practices and get the same results we have now for CCGs (and
+any higher level orgs).
+"""
+from django.core.management.base import BaseCommand
+from django.db.models import Func
+from django.contrib.gis.db.models import Collect
+
+from frontend.models import Practice, PCT
+
+
+class Command(BaseCommand):
+
+    help = __doc__
+
+    def handle(self, *args, **options):
+        infer_practice_boundaries()
+
+
+class VoronoiPolygons(Func):
+    function = 'ST_VoronoiPolygons'
+
+
+def infer_practice_boundaries():
+    ccgs = PCT.objects.filter(
+        org_type='CCG',
+        close_date__isnull=True,
+        boundary__isnull=False
+    )
+    for ccg in ccgs:
+        infer_practice_boundaries_for_ccg(ccg)
+
+
+def infer_practice_boundaries_for_ccg(ccg):
+    assert ccg.boundary is not None
+    practices = (
+        ccg.practice_set
+        .filter(location__isnull=False, setting=4)
+        .exclude(
+            status_code__in=(
+                Practice.STATUS_RETIRED,
+                Practice.STATUS_DORMANT,
+                Practice.STATUS_CLOSED
+            )
+        )
+    )
+    partition = practices.aggregate(v=VoronoiPolygons(Collect('location')))['v']
+    for region in partition:
+        # Clip the region to the CCG boundary so that practice boundaries never
+        # extend outside their CCG
+        boundary = region.intersection(ccg.boundary)
+        # Find the practices contained within the region (there may be multiple
+        # as multiple practice codes sometimes share the same location)
+        for practice in practices:
+            # We check containment using the region rather than the clipped
+            # boundary as we sometimes have practices located just outside
+            # their CCG boundary (presumably due to lack of precision in the
+            # geo data)
```

Comment:
You can say that again: https://github.com/ebmdatalab/openprescribing/issues/528